nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```diff
@@ -0,0 +1,142 @@
+            typArgument.getAnnotationMirrors();
+        boolean hasNullableAnnotation =
+            Nullness.hasNullableAnnotation(annotationMirrors.stream(), config);
+        if (hasNullableAnnotation) {
+          nullableTypeArguments.add(index);
+        }
+        index++;
+      }
+
+      Type baseType = type.tsym.type;
+      com.sun.tools.javac.util.List<Type> baseTypeArguments = baseType.getTypeArguments();
+      index = 0;
+      for (Type baseTypeArg : baseTypeArguments) {
+
+        // if type argument at current index has @Nullable annotation base type argument at that
+        // index
+        // should also have a @Nullable annotation on its upper bound.
+        if (nullableTypeArguments.contains(index)) {
+
+          Type upperBound = baseTypeArg.getUpperBound();
+          com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrors =
+              upperBound.getAnnotationMirrors();
+          boolean hasNullableAnnotation =
+              Nullness.hasNullableAnnotation(annotationMirrors.stream(), config);
+          // if base type argument does not have @Nullable annotation then the instantiation is
+          // invalid
+          if (!hasNullableAnnotation) {
+            invalidInstantiationError(tree, state, analysis);
+          }
+        }
+        index++;
+      }
+      // Generics check for nested type parameters
+      for (Type typeArgument : typeArguments) {
+        if (typeArgument.getTypeArguments().length() > 0) {
+          checkInstantiatedType(typeArgument, state, tree, analysis, config);
+        }
+      }
+    }
+  }
+
+  /** Generics checks for parameterized typed trees * */
+  public static void checkInstantiationForParameterizedTypedTree(
+      ParameterizedTypeTree tree, VisitorState state, NullAway analysis, Config config) {
+    if (!config.isJSpecifyMode()) {
+      throw new IllegalStateException("should only check type instantiations in JSpecify mode");
+    }
+    List<? extends Tree> typeArguments = tree.getTypeArguments();
+    HashSet<Integer> nullableTypeArguments = new HashSet<Integer>();
+    JCTree.JCTypeApply baseTypeTree = (JCTree.JCTypeApply) tree;
+    Type t = baseTypeTree.type;
+    Type.ClassType classType = (Type.ClassType) t;
+    // If none of the arguments are @Nullable annotated MetaData is empty
+    if (classType.tsym.getMetadata() != null) {
+      List<Attribute.TypeCompound> baseTypeAttributes =
+          classType.tsym.getMetadata().getTypeAttributes();
+      // Store the arguments in the base type that have @Nullable annotation in the set
+      for (int i = 0; i < baseTypeAttributes.size(); i++) {
+        // position - index of the parameters in the list of base type attributes that have
+        // @Nullable annotation
+        if (baseTypeAttributes.get(i).toString().equals("@org.jspecify.nullness.Nullable")) {
+          nullableTypeArguments.add(baseTypeAttributes.get(i).position.parameter_index);
+        }
+      }
+    }
```

Comment:
Is this the code that is checking for `@Nullable` on the upper bounds of the base type variables?  Which exact API calls gets the annotations on upper bounds?

Reply:
Yes. The MetaData returns the results considering the upper bounds